### PR TITLE
Add ability to re-export parity-scale-codec crate

### LINKED ISSUE #323: Feature request: ability to reexport codec crate
Hi everyone!

I found out by an accident that parity-scale-codec can't be reexport from wrapper crate. Here what I mean (repo with sample https://github.com/gshep/scale_reexport):

let there be a create that does nothing except depends on codec and reexports it:
Cargo.toml = 
```
[package]
name = "scale_reexport"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
codec = { package = "parity-scale-codec", version = "2.3.1", default-features = false, features = ["derive"] }
```
and lib.rs:
```
pub use codec;
```

And we want to use this crate in the end binary but don't want to duplicate dependency on parity-scale-codec:
Cargo.toml = 
```
[package]
name = "use_scale_reexport"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
scale_reexport = { path = "../scale_reexport" }
```

main.rs = 
```
use scale_reexport::codec::{Decode, Encode};

#[derive(Decode, Encode)]
enum TestEnum {
    FirstVariant,
    SecondVariant,
}

fn main() {
    let x = TestEnum::FirstVariant;
    let _y = x.encode();
}
```

This code is not compiled.

I was digging a bit and found `codec(crate` attribute (there is similiar for `serde`) but it is only for `MaxEncodedLen`. I was able to hack it so the following will work:
```
#[derive(Decode, Encode)]
#[codec(crate = scale_reexport::codec)]
enum TestEnum {
// ...
```

I can create corresponding PR however I would like to ask first. What branch should I checkout for 2.3.1 version, master? I see there 3.2 and 4.x branches for corresponding versions but I'm a bit confused how to do it for 2.3.1.

Thanks in advance!

### PR DESCRIPTION
**Pull request by gshep**
Closes #323

### CODE REVIEW

File: derive/src/lib.rs [hunk 1]
```diff
@@ -273,8 +246,8 @@ pub fn compact_as_derive(input: proc_macro::TokenStream) -> proc_macro::TokenStr
 		return e.to_compile_error().into();
 	}
 
-	let crate_ident = match crate::parity_scale_codec_ident() {
-		Ok(crate_ident) => crate_ident,
+	let crate_path = match codec_crate_path(&input.attrs) {
+		Ok(crate_path) => parse_quote!(#crate_path),
```
[the remark this answers is not in the document]
Comment (bkchr) [reply derive/src/lib.rs:250]: Now I see why. I think we should change the `trait_bounds::add` function to directly take a path.

[the remark this answers is not in the document]
Comment (gshep) [reply derive/src/lib.rs:250]: I found out that `encode/decode::quote` also requires `TokenStream`. So it requires a bit more changes. I'm not sure if this is a good direction

**Comment by gshep** [comment]
@bkchr Could you take a look one more time?

CI failed with the error: `error: linker `clang-13` not found`

File: derive/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bkchr) [reply derive/src/lib.rs:250]: Yes, please do it everywhere.

[the remark this answers is not in the document]
Comment (gshep) [reply derive/src/lib.rs]: done

**Comment by bkchr** [comment]
@gshep can you merge master to make the CI happy?

**Comment by gshep** [comment]
> @gshep can you merge master to make the CI happy?

Unfortunately this didn't help. CI fails now with the following:
```
Fresh parity-scale-codec v3.1.0 (/builds/parity/parity-scale-codec)
    Finished test [unoptimized + debuginfo] target(s) in 23.26s
     Running `/ci-cache/parity-scale-codec/targets/325/test-rust-stable-no_derive/debug/deps/parity_scale_codec-c67a3a418efa0b13`
/ci-cache/parity-scale-codec/targets/325/test-rust-stable-no_derive/debug/deps/parity_scale_codec-c67a3a418efa0b13:
 /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.29' not found (required by /ci-cache/parity-scale-codec/targets/325/test-rust-stable-no_derive/debug/deps/parity_scale_codec-c67a3a418efa0b13)
error: test failed, to rerun pass '--lib'
```

**Comment by bkchr** [comment]
@alvicsam any idea what is happening here? How can it happen that we don't find the glibc we are building with?

**Comment by alvicsam** [comment]
Cleaned CI cache, rerun the job - now CI is happy

**Comment by gshep** [comment]
@bkchr What should be done next? Tags `parity-scale-codec(-derive)-v2.3.1`? What about to publish this version to crates.io?

**Comment by bkchr** [comment]
Sorry for the delay. I published them :)

**Comment by HCastano** [comment]
@gshep FYI, this broke our build of ink!: https://github.com/paritytech/ink/pull/1180 😅

**Comment by gshep** [comment]
> @gshep FYI, this broke our build of ink!: [paritytech/ink#1180](https://github.com/paritytech/ink/pull/1180) sweat_smile

ah, shame on me =) .

Frankly speaking there is one more that could make some issues - https://github.com/paritytech/scale-info/pull/145

**Pull request merged**